### PR TITLE
EFS functional testing

### PR DESCRIPTION
**Pull request by sparrc**
### Summary

Adds a functional test testing writing/reading from EFS volumes mounted via Docker's NFS integration.

### Implementation details

In this test we will setup an EFS volume with mount target. We will then mount this into two tasks: first a task that writes to the EFS volume, then a task that reads from the EFS volume.

Both tasks exit with code `42` if they are successful, so we check for this exit code to verify that the test passed.

Currently this is implemented using the existing `dockerVolumeConfiguration` model. When the feature is ready this will be changed to the `EFSVolumeConfiguration` model.

### Description for the changelog

EFS volume configuration functional testing

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 1]
```diff
@@ -1359,6 +1360,106 @@ func TestASMSecretsARN(t *testing.T) {
...
+//   1. creates an EFS filesystem with a mount target.
+//   2. spins up a task to mount and write to the filesystem.
+//   3. spins up a task to mount and read from the filesystem.
+//   4. TODO: do this via EFSVolumeConfiguration instead of via NFS/Docker.
+func TestRunEFSVolumeTask(t *testing.T) {
+	if os.Getenv("TEST_DISABLE_EXECUTION_ROLE") == "true" {
+		t.Skip("TEST_DISABLE_EXECUTION_ROLE was set to true")
+	}
+
+	if IsCNPartition() {
```
Comment (sharanyad) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: why is this needed?

Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: this URL will not work in the CN partition: `addr=FILESYSTEM_ID.efs.TEST_REGION.amazonaws.com`, there is currently a TODO in the feature branch (and design doc) to deal with this in EFSVolumeConfiguration (see https://github.com/aws/amazon-ecs-agent/pull/2234/files#diff-a617e6edf02b5af84b984ccb06ce315fR547)

File: agent/Gopkg.toml [hunk 2]
```diff
@@ -19,7 +19,7 @@
 #  name = "github.com/x/y"
 #  version = "2.4.0"
 
-required = ["github.com/golang/mock/mockgen/model"]
+required = ["github.com/golang/mock/mockgen/model", "github.com/aws/aws-sdk-go/service/efs"]
```
Comment (adnxn) [inline agent/Gopkg.toml]: why is this separate from the regular aws-sdk-go constraint?

Comment (sparrc) [reply agent/Gopkg.toml]: Yes I'm going to figure out how to resolve this. The reason was that we are only using the efs client in a linux-specific unit test, so I wasn't able to get dep to recognize the dependency from my work laptop.

Comment (sparrc) [reply agent/Gopkg.toml]: this is resolved now, deleting the entry and running "dep ensure" again seems to have fixed it

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 3]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
...
+	creationToken := "efs-func-tests"
+	_, err := efsClient.CreateFileSystem(&efs.CreateFileSystemInput{
+		CreationToken: aws.String(creationToken),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case efs.ErrCodeFileSystemAlreadyExists:
+				t.Logf("EFS filesystem already exists")
```
Comment (sharanyad) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Do we expect to reuse the same file system created once for the tests? I don't see any cleanup logic.

Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: yes, we expect to reuse the filesystem, that's why we check for ErrCodeFileSystemAlreadyExists (filesystem already exists) and ErrCodeMountTargetConflict (mount target already exists)

The main reason for this is speed, and the EFS api makes it easy with the CreationToken: https://docs.aws.amazon.com/efs/latest/ug/API_CreateFileSystem.html#API_CreateFileSystem_RequestSyntax

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 4]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
...
+//   2. spins up a task to mount and write to the filesystem.
+//   3. spins up a task to mount and read from the filesystem.
+//   4. TODO: do this via EFSVolumeConfiguration instead of via NFS/Docker.
+func TestRunEFSVolumeTask(t *testing.T) {
+	if os.Getenv("TEST_DISABLE_EXECUTION_ROLE") == "true" {
+		t.Skip("TEST_DISABLE_EXECUTION_ROLE was set to true")
+	}
+
+	if IsCNPartition() {
+		t.Skip("Skip TestSSMSecretsEncryptedParameter in China partition")
```
Comment (fenxiong) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: TestSSMSecretsEncryptedParameter -> TestRunEFSVolumeTask

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 5]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
 	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42; got %d", exitCode))
 }
 
+// TestRunEFSVolumeTask does the following:
+//   1. creates an EFS filesystem with a mount target.
+//   2. spins up a task to mount and write to the filesystem.
+//   3. spins up a task to mount and read from the filesystem.
+//   4. TODO: do this via EFSVolumeConfiguration instead of via NFS/Docker.
+func TestRunEFSVolumeTask(t *testing.T) {
+	if os.Getenv("TEST_DISABLE_EXECUTION_ROLE") == "true" {
```
Comment (fenxiong) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: why is this check needed?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 6]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
...
+		t.Skip("TEST_DISABLE_EXECUTION_ROLE was set to true")
+	}
+
+	if IsCNPartition() {
+		t.Skip("Skip TestSSMSecretsEncryptedParameter in China partition")
+	}
+
```
Comment (fenxiong) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: i think you should have the following to let agent runs as host network mode (see https://github.com/aws/amazon-ecs-agent/issues/1873)
```
os.Setenv("ECS_FTEST_FORCE_NET_HOST", "true")
defer os.Unsetenv("ECS_FTEST_FORCE_NET_HOST")
```

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 4 above)
Comment (petderek) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: update this to say EFS :)

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 7]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
...
+		"TEST_REGION":   *ECS.Config.Region,
+	}
+	wTask, err := agent.StartTaskWithTaskDefinitionOverrides(t, "task-efs-vol-write", overrides)
+	require.NoError(t, err, "Register task definition failed")
+	// Wait for the first task to create the volume
+	wErr := wTask.WaitStopped(waitTaskStateChangeDuration)
+	require.NoError(t, wErr, "Error waiting for task to transition to STOPPED")
+	wExitCode, ok := wTask.ContainerExitcode("task-efs-vol-write")
+	require.True(t, ok, "Error code for container [task-efs-vol-write] not found, check the logs")
+	assert.Equal(t, 42, wExitCode, fmt.Sprintf("Expected exit code of 42; got %d", wExitCode))
```
Comment (fenxiong) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: nit: i think this can just be require since there's no need to continue if this fails?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 8]
```diff
@@ -1359,6 +1360,126 @@ func TestASMSecretsARN(t *testing.T) {
...
+				require.NoError(t, err, "Unexpected error from DescribeFileSystems: %s", err)
+				return *out.FileSystems[0].FileSystemId
+			default:
+				require.NoError(t, err, "Error creating EFS Filesystem")
+			}
+		}
+	}
+
+	// Wait for filesystem to be "available"
+	for true {
```
Comment (petderek) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: I'd prefer if this had a finite timeout -- like try 10 times or so.

File: agent/functional_tests/util/utils.go [hunk 9]
```diff
@@ -809,8 +809,8 @@ func AttributesToMap(attributes []*ecs.Attribute) map[string]string {
 	return attributeMap
 }
 
-// getSubnetID gets the subnet id for the instance from ec2 instance metadata
-func getSubnetID() (string, error) {
+// GetSubnetID gets the subnet id for the instance from ec2 instance metadata
+func GetSubnetID() (string, error) {
```
Comment (petderek) [inline agent/functional_tests/util/utils.go]: nice

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 4 above)
Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: will do

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 8 above)
Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: okay, will fix that

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 7 above)
Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: 👍

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 6 above)
Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: 👍

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (sparrc) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: That's probably a c&p artifact, I'll remove it

**Comment by yumex93** [comment]
LGTM. Don't forget to squash your commits before merging to dev.
